pkg/storage/chunk/cache/memcached.go

```diff
@@ -266,13 +286,10 @@ func (c *Memcached) Stop() {
 // Assumes c.inputCh, c.closed channels are non-nil
 // Go routine safe and idempotent.
 func (c *Memcached) closeAndStop() {
-	c.stopMu.Lock()
-	defer c.stopMu.Unlock()
-
-	if !c.stopped {
+	if !c.stopped.Load() {
```

Comment:
I think you should use https://pkg.go.dev/sync/atomic#Bool.CompareAndSwap here for increased safety and protection against races.

Reply:
I'm bit confused to use `CompareAndSwap` for bool here.

The expected behaviour I want is I want some channels to be closed after Loading and before Swapping. which is not possible with single `CompareAndSwap`?

If I use `CompareAndSwap()` instead of `Load` here, then bool may be already updated even before closing this channels. 

The more I think about it, mutex instead of atomic.Bool would be right type here. wdyt?

Reply:
Yeah you're right.
The more I look at this, the more I think this should be a `sync.Once`, actually.